narwhal/primary/src/aggregators.rs

```diff
@@ -17,14 +19,18 @@ pub struct VotesAggregator {
     weight: Stake,
     votes: Vec<(PublicKey, Signature)>,
     used: HashSet<PublicKey>,
+    metrics: Arc<PrimaryMetrics>,
 }
 
 impl VotesAggregator {
-    pub fn new() -> Self {
+    pub fn new(metrics: Arc<PrimaryMetrics>) -> Self {
+        metrics.votes_received_last_round.set(0);
```

Comment:
Good point @aschran , and to be honest I am ok with it as we want to use it as a rough signal of the vote stuff. As I explained to @arun-koshy I am mostly see value on this metric when ever we get stalled - and that's when I expect the latest votes snapshot to win here, which should be the latest header, but as you pointed out we don't have strong guarantees

Reply:
If you don't want to fix the thread safety (which could be a valid time-saving decision) then maybe add `unreliable_` or something like that to the metric name?

Reply:
Actually thought a little bit better about this. I believe we do cancel the previously created header handler before we go into proposing the new one (our earlier discussion triggered that thought - otherwise that would be a great opportunity for improvement). More specifically we do this here https://github.com/MystenLabs/sui/blob/ad91a6afd7b01e1d5a9e50cd922d046247f31754/narwhal/primary/src/core.rs#L708 . So I would expect even on the previous future on the next cycle to either:
* have the process cancelled
* or the certificate formed already so there is no overlap anymore

so it seems to me that we are ok. What do you think @aschran ?

Reply:
The cancellation is not synchronous, it is received in the vote request loop of `propose_header` and the cancellation arm of the select is not guaranteed to ever run. Unfortunately I don't think this best-effort cancellation makes the metric as you have it here thread-safe.

Reply:
I like the proposal (1) @mwtian . I believe this will even give us stronger guarantees that we don't have orphaned tasks running in the background until completion - which might even be out of our hand at some point. Although I can see cases where this might introduce some latency until cancelling previous feature, I think it would be the cleanest of all solutions. 

As this can affect other parts of the protocol, I'll go ahead and merge this PR , and open a new one to tackle this specific case.

Reply:
You may or may not want to do (1) for other reasons, but I don't think it's a good solution to the problem of the thread-unsafety of this metric. Even if you remove the current place where multiple `VoteAggregator` instances may exist, there is nothing in the `VoteAggregator` struct itself to prevent multiple from existing simultaneously or suggest that it shouldn't be done. Future changes could easily reintroduce the problem.

I think the best solution would be what I suggested over chat @akichidis:
Add a Drop handler to VoteAggregator that instead of resetting to zero, when the VoteAggregator is dropped, decrements the metric by the # of votes. Then you might be counting votes from multiple pending headers at once but at least the number of "outstanding" votes would always be accurate. And if it gets stuck somewhere there almost certainly will be only one header pending at the relevant time.

If you don't like that solution then I think you should change the way `VotesAggregator` are constructed to make it impossible to have more than one at a time.